### ISSUE TITLE
S1854 false positive for out parameters

### ISSUE THREAD

**Issue by jmarolf** [SonarSource/sonar-csharp#689]
### Description

[S1854](http://www.sonarlint.org/visualstudio/rules/index.html#version=6.3.0.2862&ruleId=S1854) does not consider assignment as part of an `out` parameter

### Repro steps

1. Crate code like the following
```C#
class Program
{
    static bool GetBool(string boolString, bool defaultValue)
    {
        bool boolValue = defaultValue; // <- S1854 triggers here
        bool.TryParse(boolString ?? "false", out boolValue);
        return boolValue;
    }
}

```

### Expected behavior

If a variable is passed as part of an `out` parameter it should no be considered a dead store.

### Actual behavior

false positive.

* SonarC# Version: 6.3.0.2862
* Visual Studio Version: 2017

**Comment by valhristov** [comment]
You are obliged to set the out variable from within the method (`TryParse` in this case), hence `bool boolValue = defaultValue` is considered dead store - it is always overwritten and never used. Better way to do this is:

```
static bool GetBool(string boolString, bool defaultValue)
{
    bool boolValue;
    return bool.TryParse(boolString ?? "false", out boolValue)
        ? boolValue
        : defaultValue;
}
```
